source/extensions/common/wasm/wasm_vm_base.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include "envoy/stats/scope.h"
+#include "envoy/stats/stats.h"
+#include "envoy/stats/stats_macros.h"
+
+#include "extensions/common/wasm/wasm_vm.h"
+
+#include "absl/strings/str_cat.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Wasm {
+
+/**
+ * Wasm host stats.
+ */
+#define ALL_VM_STATS(COUNTER, GAUGE)                                                               \
+  COUNTER(created)                                                                                 \
+  COUNTER(cloned)                                                                                  \
+  GAUGE(active, NeverImport)
+
+struct VmStats {
+  ALL_VM_STATS(GENERATE_COUNTER_STRUCT, GENERATE_GAUGE_STRUCT)
+};
+
+struct VmGlobalStats {
+  std::atomic<int64_t> active_;
+};
+
+// Wasm VM base instance. Provides common behavior (e.g. Stats).
+class WasmVmBase : public WasmVm {
+public:
+  WasmVmBase(Stats::ScopeSharedPtr scope, VmGlobalStats* global_stats_ptr,
+             absl::string_view runtime)
+      : scope_(scope), global_stats_ptr_(global_stats_ptr),
+        stats_(VmStats{
+            ALL_VM_STATS(POOL_COUNTER_PREFIX(*scope_, absl::StrCat("wasm_vm.", runtime, ".")),
+                         POOL_GAUGE_PREFIX(*scope_, absl::StrCat("wasm_vm.", runtime, ".")))}),
+        runtime_(std::string(runtime)) {
+    global_stats_ptr_->active_++;
+    stats_.created_.inc();
+    stats_.active_.set(global_stats_ptr_->active_);
+    ENVOY_LOG(debug, "WasmVm created {} now active", runtime_, global_stats_ptr_->active_);
+  }
+  virtual ~WasmVmBase() {
+    global_stats_ptr_->active_--;
+    stats_.active_.set(global_stats_ptr_->active_);
```

Comment:
aren't we about to destruct stats_ and thus stats_.active_?

I suspect you are doing this because in the stats infrastructure this gets tied by name, potentially, to other instances of the same named stat. Doc that subtlety.

Reply:
Done.  Perhaps there is a better way of doing this?